src/Ide/Plugin/Eval.hs

```diff
@@ -0,0 +1,299 @@
+  where
+    checkMatch = T.stripPrefix "-- >>> "
+    looksLikeSplice l
+      | Just l' <- T.stripPrefix "--" l
+      = not (" >>>" `T.isPrefixOf` l')
+      | otherwise
+      = False
+
+    goSearch _ [] = []
+    goSearch line (l : ll)
+      | Just match <- checkMatch l =
+        goAcc (line + 1) [(match, line)] ll
+      | otherwise =
+        goSearch (line + 1) ll
+
+    goAcc line acc [] = [(reverse acc,Range p p)] where p = Position line 0
+    goAcc line acc (l:ll)
+      | Just match <- checkMatch l =
+        goAcc (line + 1) ([(match, line)] <> acc) ll
+      | otherwise =
+        (reverse acc,r) : goSearch (line + 1) ll
+      where
+        r = Range p p'
+        p = Position line 0
+        p' = Position (line + spliceLength) 0
+        spliceLength = length (takeWhile looksLikeSplice (l:ll))
+
+provider :: CodeLensProvider
+provider lsp _state plId CodeLensParams {_textDocument} = response $ do
+  let TextDocumentIdentifier uri = _textDocument
+  contents <- liftIO $ getVirtualFileFunc lsp $ toNormalizedUri uri
+  let text = Rope.toText . (_text :: VirtualFile -> Rope.Rope) <$> contents
+  let matches = extractMatches text
+
+  cmd <- liftIO $ mkLspCommand plId evalCommandName "Evaluate..." (Just [])
+
+  let lenses =
+        [ CodeLens range (Just cmd') Nothing
+          | (m, r) <- matches,
+            let (_, startLine) = head m
+                (_, endLine) = last m
+                range = Range start end
+                start = Position startLine 0
+                end = Position endLine 1000
+                args = EvalParams m r _textDocument,
+            let cmd' = (cmd :: Command)
+                    {_arguments = Just (List [toJSON args])
+                    ,_title = if trivial r then "Evaluate..." else "Refresh..."
+                    }
+        ]
+
+  return $ List lenses
+  where
+      trivial (Range p p') = p == p'
+
+evalCommandName :: CommandId
+evalCommandName = "evalCommand"
+
+evalCommand :: PluginCommand
+evalCommand =
+  PluginCommand evalCommandName "evaluate" runEvalCmd
+
+data EvalParams = EvalParams
+  { statements :: [(Text, Int)],
+    editTarget :: !Range,
+    module_    :: !TextDocumentIdentifier
+  }
+  deriving (Eq, Show, Generic, FromJSON, ToJSON)
+
+runEvalCmd :: CommandFunction EvalParams
+runEvalCmd lsp state EvalParams {..} = response' $ do
+  let TextDocumentIdentifier {_uri} = module_
+  fp <- handleMaybe "uri" $ uriToFilePath' _uri
+  contents <- liftIO $ getVirtualFileFunc lsp $ toNormalizedUri _uri
+  text <- handleMaybe "contents" $ Rope.toText . (_text :: VirtualFile -> Rope.Rope) <$> contents
+
+  session <-
+    liftIO
+      $ runAction "runEvalCmd.ghcSession" state
+      $ use_ GhcSessionDeps
+      $ toNormalizedFilePath'
+      $ fp
+
+  ms <-
+    liftIO
+      $ runAction "runEvalCmd.getModSummary" state
+      $ use_ GetModSummary
+      $ toNormalizedFilePath'
+      $ fp
+
+  now <- liftIO getCurrentTime
+
+  (temp, clean) <- liftIO newTempFile
```

Comment:
Why not withTempFile/withTempFile/withFile - if you have an exception after opening the file, but before the finally, you won't close it down properly.

Reply:
I wanted to use `withTempFile`, but it has an unnecessarily restrictive type. This do block is in the `ExceptT IO` monad. If `withTempFile` had the more general type below, I would have been able to use it.
```
withTempFile :: MonadIO m => (FilePath -> m a) -> m a 
```

Reply:
Such a type is not possible because of the possibility of exceptions. You need a Monad that at least support `bracket`-like exception catcher (which cannot able to be lifted by `liftIO`)

Reply:
I believe the `exceptions` package provides a generalized `bracket`, doesn't it?

Reply:
Oh, of course if there are extra constraints it's possible to define it more generally. I didn't mean that it should be restricted to IO monad. If I confused you, I'm sorry.

Reply:
Yep, you're right. This is a reasonable way to go for now (not ideal, but the cost of a leaked temp file very occasionally isn't too high)

Reply:
I've switched to the `temporary` package which provides a more polymorphic `with..` function that works with the `ExceptT IO` monad